### PR TITLE
Test fusing two computes: Matmul + Unary Sfpu

### PR DESCRIPTION
**Pull request by skotaracTT**
### Ticket

https://github.com/orgs/tenstorrent/projects/123/views/1?pane=issue&itemId=109135371&issue=tenstorrent%7Ctt-llk%7C160
### Problem description

Adding test that performs a fused operation consisting of a matrix multiplication followed by an Sfpu unary op. Two input tensors are multiplied using matmul, and the resulting tensor is fed as input for the Sfpu op.
### What's changed

The fusion is implemented via two sequential LLK pipeline runs:
- The first run computes the matmul. The second run applies the Sfpu unary op on the matmul result.
- Synchronization: Semaphores used to ensure correct ordering between kernels—specifically, the second unpacker must wait for the first packer to complete writing the matmul result to L1, before unpacker kernel reads it.
- Kernel Setup: Kernel inits/states are reconfigured between runs to accommodate the transition from matmul to Sfpu op execution.
### Note
There is percision loss when doing two consecutive computes, and certain combinations must be skipped until further investigation.
- Will add issue to address further investigation as to why certain combinations of **Math Fidelity + Format + Sfpu Op** are causing loss in precision when running two computes (matmul + sfpu op)
### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Adding Test to LLK Test Infra
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### CODE REVIEW

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        MathOperation.Cos: lambda x: math.cos(x),
+        MathOperation.Log: lambda x: math.log(x) if x != 0 else float("nan"),
+        MathOperation.Reciprocal: lambda x: 1 / x if x != 0 else float("nan"),
+        MathOperation.Sin: lambda x: math.sin(x),
+        MathOperation.Sqrt: lambda x: math.sqrt(x),
+        MathOperation.Square: lambda x: x * x,
+    }
+    if operation not in ops:
+        raise ValueError("Unsupported operation!")
+    # return [ops[operation](num) for num in tensor1_float.tolist()][:256]
```
Comment (fvranicTT) [inline tests/python_tests/test_matmul_and_unary_sfpu.py:73]:
```suggestion
```

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        pytest.skip(reason="This combination is not fully implemented in testing")
+    if formats.input_format == DataFormat.Float16_b and mathop in [
+        MathOperation.Sin,
+        MathOperation.Cos,
+    ]:
+        pytest.skip(reason="This combination is not fully implemented in testing")
+
+    if (
+        formats.input_format == DataFormat.Float16
+        and DestAccumulation.Yes
```
Comment (fvranicTT) [inline tests/python_tests/test_matmul_and_unary_sfpu.py:149]: This looks odd, we're missing variable comparison here (i.e. `dest_acc == DestAccumulation.Yes`) and fidelity too (i.e. `math_fidelity == MathFidelity.LoFi`)

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        and MathFidelity.LoFi
+        and mathop in [MathOperation.Sin, MathOperation.Cos, MathOperation.Square]
+    ):
+        pytest.skip(reason="This combination is not fully implemented in testing")
+    if (
+        formats.input_format == DataFormat.Float16_b
+        and MathFidelity.LoFi
```
Comment (fvranicTT) [inline tests/python_tests/test_matmul_and_unary_sfpu.py:156]: ditto

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        and mathop == MathOperation.Square
+    ):
+        pytest.skip(reason="This combination is not fully implemented in testing")
```
Comment (fvranicTT) [inline tests/python_tests/test_matmul_and_unary_sfpu.py:159]: If I'm reading this correctly, all of these skip checks could be merged into something like:
```python
if formats.input_format == DataFormat.Float16:
    if (
        (dest_acc == DestAccumulation.No and get_chip_architecture() == ChipArchitecture.BLACKHOLE)
        or (dest_acc == DestAccumulation.Yes and math_fidelity == MathFidelity.LoFi and mathop in [MathOperation.Sin, MathOperation.Cos, MathOperation.Square])
    ):
        pytest.skip(reason="This combination is not fully implemented in testing")
elif formats.input_format == DataFormat.Float16_b:
    if (
        (mathop in [MathOperation.Sin, MathOperation.Cos])
        or (math_fidelity == MathFidelity.LoFi and mathop == MathOperation.Square)
    ):
        pytest.skip(reason="This combination is not fully implemented in testing")
```

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    generate_params,
+    input_output_formats,
+)
+from helpers.stimuli_generator import generate_stimuli
+from helpers.test_config import generate_make_command
+from helpers.tilize_untilize import tilize
+from helpers.utils import compare_pcc, run_shell_command
+
+
+def generate_matmul_golden(operand1, operand2, data_format, math_fidelity):
```
Comment (fvranicTT) [inline tests/python_tests/test_matmul_and_unary_sfpu.py:36]: We should consider moving golden generation to a separate class, to keep the code [dry](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). We're now adding this method for the 3rd time. Consider this a future improvement, no need to do it as part of this pull request.

File: tests/sources/matmul_and_unary_sfpu_test.cpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdio>
+
+#include "ckernel.h"
+#include "llk_defs.h"
+
+// Globals
+uint32_t unp_cfg_context          = 0;
```
[the remark this answers is not in the document]
Comment (fvranicTT) [reply tests/sources/matmul_and_unary_sfpu_test.cpp:13]: Actually, we need a number of global variables (😞 ), and it could very well be that the code won't compile if this suggestion is applied. But we could give it a shot.

File: tests/sources/matmul_and_unary_sfpu_test.cpp [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+    t6_semaphore_post<>(semaphore::PACK_DONE);
+
+    // Start of second pack kernel to perform final pack after executing matmul on tilized data
+    _llk_pack_reconfig_data_format_<is_fp32_dest_acc_en>(PACK_IN, PACK_OUT, tile_size);
+#ifdef PACK_DST_BFP8_B
+    constexpr auto PACK_OUT = static_cast<std::underlying_type_t<DataFormat>>(DataFormat::Float16_b);
+#endif
+
+    _llk_pack_init_<false, false, DstTileFaceLayout::RowMajor, false>(PACK_OUT);
```
[the remark this answers is not in the document]
Comment (fvranicTT) [reply tests/sources/matmul_and_unary_sfpu_test.cpp:168]: This suggestion seems to be invalid.

File: tests/sources/matmul_and_unary_sfpu_test.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    _llk_math_matmul_<MATH_FIDELITY, DstTileFaceLayout::RowMajor>(0);
+    _llk_math_dest_section_done_<DstSync::SyncHalf, is_fp32_dest_acc_en>();
+
+    // Start of second math kernel to perform matmul on now tilized input data
+    _llk_math_reconfig_data_format_srca_<false, is_fp32_dest_acc_en>(MATH_FORMAT);
+    // copy srca to dest
+#ifdef ARCH_BLACKHOLE
+    _llk_math_eltwise_unary_datacopy_init_<DataCopyType::A2D, BroadcastType::NONE, false, is_fp32_dest_acc_en, false>(0, 0, 4, MATH_FORMAT);
+#else
+    _llk_math_eltwise_unary_datacopy_init_<DataCopyType::A2D, BroadcastType::NONE, is_fp32_dest_acc_en, false>(0, 0, 4, MATH_FORMAT);
```
[the remark this answers is not in the document]
Comment (fvranicTT) [reply tests/sources/matmul_and_unary_sfpu_test.cpp:114]: This suggestion seems to be invalid, BH and WH have a different type of template arguments.

File: tests/python_tests/test_matmul_and_unary_sfpu.py (hunk 2 above)
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:148]: Good catch!

File: tests/python_tests/test_matmul_and_unary_sfpu.py (hunk 3 above)
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:155]: I don't know how this slipped 😅

File: tests/python_tests/test_matmul_and_unary_sfpu.py (hunk 5 above)
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:36]: Yes I agree. I was thinking about opening an issue about this.🤝

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    result_matrix = torch.matmul(operand1_matrix, operand2_matrix)
+
+    return result_matrix.view(1024).to(format_dict[data_format])
+
+
+def generate_sfpu_golden(operation, operand1, data_format):
+    tensor1_float = (
+        operand1.clone()
+        .detach()
+        .to(format_dict[data_format] if data_format != "Bfp8_b" else torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:60]: Agreed

File: tests/python_tests/test_matmul_and_unary_sfpu.py (hunk 4 above)
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:158]: Works!

File: tests/python_tests/test_matmul_and_unary_sfpu.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:60]: I think that I'll make a variable `torch_format = format_dict.get(data_format, torch.bfloat16)` because this is used in multiple places, and could help readability and avoid repetition.

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+)
+param_ids = generate_param_ids(all_params)
+
+
+@pytest.mark.parametrize(
+    "testname, formats, dest_acc, approx_mode, mathop, math_fidelity",
+    clean_params(all_params),
+    ids=param_ids,
+)
+def test_matmul(testname, formats, dest_acc, approx_mode, mathop, math_fidelity):
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:134]: Fixed!

**Review by fvranicTT — APPROVED** [review]

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    result_matrix = torch.matmul(operand1_matrix, operand2_matrix)
+
+    return result_matrix.view(1024).to(data_type)
+
+
+def generate_sfpu_golden(operation, operand1, data_format):
+    data_type = format_dict.get(data_format, format_dict[DataFormat.Float16_b])
+    tensor1_float = operand1.clone().detach().to(data_type)
+    ops = {
+        MathOperation.Abs: lambda x: abs(x),
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:59]: abs is a built-in function, which returns an int or float depending on the input, whereas math.abs does not exist. If I try to use it, l get an error. There is math.fabs which returns always returns a float.

[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:59]: abs is a built-in function, which returns an int or float depending on the input, whereas math.abs does not exist. If I try to use it l get an error. There is math.fabs which  always returns a float.

File: tests/python_tests/test_matmul_and_unary_sfpu.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        MathFidelity.HiFi2,
+        MathFidelity.HiFi3,
+        MathFidelity.HiFi4,
+    ],
+)
+param_ids = generate_param_ids(all_params)
+
+
+@pytest.mark.parametrize(
+    "testname, formats, dest_acc, approx_mode, mathop, math_fidelity",
```
[the remark this answers is not in the document]
Comment (skotaracTT) [reply tests/python_tests/test_matmul_and_unary_sfpu.py:126]: I will do this in a separate PR for all the test files, just so I don't add them to this one
